File: lib/data/repositories/manual_history_repository.dart
import 'package:drift/drift.dart';

import '../../domain/models/manual_history_entry.dart';
import '../../infra/db/app_database.dart';

class ManualHistoryRepository {
  final AppDatabase _db;
  ManualHistoryRepository(this._db);

  Future<int> insert({
    required int clientId,
    required DateTime date,
    required int small,
    required int large,
    String? note,
  }) async {
    final now = DateTime.now().millisecondsSinceEpoch;
    return _db.into(_db.manualHistoryEntriesTable).insert(
          ManualHistoryEntriesTableCompanion.insert(
            clientId: clientId,
            date: _toEpochDays(date),
            sheepCountSmall: Value(small),
            sheepCountLarge: Value(large),
            note: Value(note),
            createdAt: now,
            updatedAt: now,
          ),
        );
  }

  Future<void> update(
    int id, {
    required DateTime date,
    required int small,
    required int large,
    String? note,
  }) async {
    await (_db.update(_db.manualHistoryEntriesTable)
          ..where((t) => t.id.equals(id)))
        .write(
      ManualHistoryEntriesTableCompanion(
        date: Value(_toEpochDays(date)),
        sheepCountSmall: Value(small),
        sheepCountLarge: Value(large),
        note: Value(note),
        updatedAt: Value(DateTime.now().millisecondsSinceEpoch),
      ),
    );
  }

  Future<void> delete(int id) async {
    await (_db.delete(_db.manualHistoryEntriesTable)
          ..where((t) => t.id.equals(id)))
        .go();
  }

  Future<List<ManualHistoryEntry>> listForClient(int clientId) async {
    final rows = await (_db.select(_db.manualHistoryEntriesTable)
          ..where((t) => t.clientId.equals(clientId))
          ..orderBy([(t) => OrderingTerm.desc(t.date)]))
        .get();
    return rows.map(_toDomain).toList();
  }

  /// Returns `(clientId, dateEpochDays)` pairs for every manual entry whose
  /// date is `>= seasonEpochDays`. Used by `ClientRepository.listAllWithStatus`
  /// to mark clients as "done this season" when they only have a manual entry.
  Future<List<({int clientId, int dateEpochDays})>>
      listClientDatesSinceEpochDays(int seasonEpochDays) async {
    final rows = await (_db.select(_db.manualHistoryEntriesTable)
          ..where((t) => t.date.isBiggerOrEqualValue(seasonEpochDays)))
        .get();
    return [
      for (final r in rows)
        (clientId: r.clientId, dateEpochDays: r.date),
    ];
  }

  static int _toEpochDays(DateTime d) {
    final utc = DateTime.utc(d.year, d.month, d.day);
    return utc.millisecondsSinceEpoch ~/ 86400000;
  }

  ManualHistoryEntry _toDomain(ManualHistoryEntryRow r) {
    final utc = DateTime.fromMillisecondsSinceEpoch(
      r.date * 86400000,
      isUtc: true,
    );
    return ManualHistoryEntry(
      id: r.id,
      clientId: r.clientId,
      date: DateTime(utc.year, utc.month, utc.day),
      small: r.sheepCountSmall,
      large: r.sheepCountLarge,
      note: r.note,
    );
  }
}

File: src/data/repositories/manual-history-repository.ts
import { desc, eq } from 'drizzle-orm';
import { manualHistoryEntries } from '@/infra/db/schema';
import type { Db } from '@/infra/db/client';
import { ManualHistoryEntry } from '@/domain/models/manual-history-entry';
import { Payment } from '@/domain/models/payment';

interface ManualHistoryRow {
  id: string;
  clientId: string;
  date: string;
  notes: string | null;
  services: string;
  paymentMethodId: string | null;
  paymentMethodLabelSnapshot: string | null;
  isPaid: number;
  paidAt: string | null;
}

function toRow(e: ManualHistoryEntry) {
  return {
    id: e.id,
    clientId: e.clientId,
    date: e.date,
    notes: e.notes,
    services: JSON.stringify(e.services),
    paymentMethodId: e.payment.methodId,
    paymentMethodLabelSnapshot: e.payment.methodLabelSnapshot,
    isPaid: e.payment.isPaid ? 1 : 0,
    paidAt: e.payment.paidAt,
  };
}

function fromRow(r: ManualHistoryRow): ManualHistoryEntry {
  return ManualHistoryEntry.parse({
    id: r.id,
    clientId: r.clientId,
    date: r.date,
    notes: r.notes,
    services: JSON.parse(r.services),
    payment: {
      methodId: r.paymentMethodId,
      methodLabelSnapshot: r.paymentMethodLabelSnapshot,
      isPaid: r.isPaid === 1,
      paidAt: r.paidAt,
    },
  });
}

export class ManualHistoryRepository {
  constructor(private readonly db: Db) {}

  async listByClient(clientId: string): Promise<ManualHistoryEntry[]> {
    const rows = await this.db
      .select()
      .from(manualHistoryEntries)
      .where(eq(manualHistoryEntries.clientId, clientId))
      .orderBy(desc(manualHistoryEntries.date));
    return rows.map((r) => fromRow(r as ManualHistoryRow));
  }

  async upsert(e: ManualHistoryEntry): Promise<void> {
    const row = toRow(e);
    await this.db
      .insert(manualHistoryEntries)
      .values(row)
      .onConflictDoUpdate({ target: manualHistoryEntries.id, set: row });
  }

  async markEntryPayment(entryId: string, payment: Payment): Promise<void> {
    await this.db.update(manualHistoryEntries).set({
      paymentMethodId: payment.methodId,
      paymentMethodLabelSnapshot: payment.methodLabelSnapshot,
      isPaid: payment.isPaid ? 1 : 0,
      paidAt: payment.paidAt,
    }).where(eq(manualHistoryEntries.id, entryId));
  }

  async delete(id: string): Promise<void> {
    await this.db.delete(manualHistoryEntries).where(eq(manualHistoryEntries.id, id));
  }
}
